clang/lib/DPCT/SaveNewFiles.cpp

```diff
@@ -174,15 +174,18 @@ bool rewriteDir(SmallString<512> &FilePath, const StringRef InRoot,
 }
 
 void rewriteFileName(SmallString<512> &FileName) {
+  const auto Extension = path::extension(FileName);
   SourceProcessType FileType = GetSourceFileType(FileName.str());
-
-  if (FileType & SPT_CudaSource) {
-    path::replace_extension(FileName, "dp.cpp");
-  } else if (FileType & SPT_CppSource) {
-    auto Extension = path::extension(FileName);
-    path::replace_extension(FileName, Extension + ".dp.cpp");
-  } else if (FileType & SPT_CudaHeader) {
-    path::replace_extension(FileName, "dp.hpp");
+  // If user does not specify which extension need be changed, we change all the
+  // SPT_CudaSource, SPT_CppSource and SPT_CudaHeader files.
+  if (DpctGlobalInfo::getChangeExtensions().empty() ||
+      DpctGlobalInfo::getChangeExtensions().count(Extension.str())) {
```

Comment:
eg. when user want to change:  a.xyz ==> a.xyz.dp.cpp,  it will not happen accord to current logic,  right? (As .xyz is not in SPT_*  defines.

Reply:
No, here is the logic in function `GetSourceFileType`:
```c++
    // clang-format off
    // For unknown file extensions, determine the file type according to:
    // A. If it shows up in the compilation database as single migration
    //    file, then treat it as a main source file.
    // B. If it is included by another source file, then treat it as a header
    //    file.
    // C. If both A and B hold, then default to A.
    // clang-format on
    auto &FileSetInDB = dpct::DpctGlobalInfo::getFileSetInCompiationDB();
    if (FileSetInDB.find(SourcePath.str()) != end(FileSetInDB)) {
      return SPT_CppSource;
    }
```
it will also return SPT_CppSource when the file is in database.